pkg/controller/deployment/deployment_controller.go

```diff
@@ -754,6 +742,158 @@ func (dc *DeploymentController) getNewReplicaSet(deployment *extensions.Deployme
 	return createdRS, dc.updateDeploymentRevision(deployment, newRevision)
 }
 
+// getNewReplicaSetAndSyncHashKey returns a replica set that matches the intent of the given deployment.
+// Returns nil if the new replica set doesn't exist yet.
+// Note that the pod-template-hash will be added to adopted RSes and pods.
+func (dc *DeploymentController) getNewReplicaSetAndSyncHashKey(deployment *extensions.Deployment, c clientset.Interface) (*extensions.ReplicaSet, error) {
+	rsList, _, err := dc.rsAndPodsWithHashKeySynced(deployment, c)
+	if err != nil {
+		return nil, fmt.Errorf("error labeling replica sets and pods with pod-template-hash: %v", err)
+	}
+	return deploymentutil.FindNewReplicaSet(deployment, rsList)
+}
+
+// getOldReplicaSetsAndSyncHashKey returns two sets of old replica sets targeted by the given Deployment; get PodList and ReplicaSetList with input functions.
+// Note that the first set of old replica sets doesn't include the ones with no pods, and the second set of old replica sets include all old replica sets.
+// Note that the pod-template-hash will be added to adopted RSes and pods.
+func (dc *DeploymentController) getOldReplicaSetsAndSyncHashKey(deployment *extensions.Deployment, c clientset.Interface) ([]*extensions.ReplicaSet, []*extensions.ReplicaSet, error) {
+	rsList, podList, err := dc.rsAndPodsWithHashKeySynced(deployment, c)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error labeling replica sets and pods with pod-template-hash: %v", err)
+	}
+	return deploymentutil.FindOldReplicaSets(deployment, rsList, podList)
+}
+
+// rsAndPodsWithHashKeySynced returns the RSes and pods the given deployment targets, with pod-template-hash information synced.
+func (dc *DeploymentController) rsAndPodsWithHashKeySynced(deployment *extensions.Deployment, c clientset.Interface) ([]extensions.ReplicaSet, *api.PodList, error) {
```

Comment:
You don't really need to pass the client as an argument since it's already a part of the controller struct.